src/Databases/DatabaseReplicated.cpp

```
@@ -810,6 +813,65 @@ void DatabaseReplicated::stopLoading()
+{
+    auto zookeeper = getZooKeeper();
+    auto local_context = getContext();
+
+    std::vector<std::string> tables_metadata_in_zk;
+    Coordination::Stat entity_name_stat;
+
+    const String metadata_node_path = zookeeper_path + "/metadata";
+
+    const auto error_code = zookeeper->tryGetChildren(metadata_node_path, tables_metadata_in_zk, &entity_name_stat);
+
+    if (error_code != Coordination::Error::ZOK)
+        throw Coordination::Exception::fromPath(error_code, metadata_node_path);
+
+    if (!tables_metadata_in_zk.empty())
+    {
+        LOG_INFO(log, "Table's metadata {} was restored from another replica", metadata_node_path);
+        return;
+    }
+
+    auto txn = std::make_shared<ZooKeeperMetadataTransaction>(zookeeper, zookeeper_path, true, "");
+    UInt64 tables_digest{};
+
+    for (auto existing_tables_it = getTablesIterator(local_context, {}, /*skip_not_loaded=*/false); existing_tables_it->isValid();
+         existing_tables_it->next())
+    {
+        const String table_name = existing_tables_it->name();
+        LOG_TEST(log, "table_name={} was found to restore metadata in zookeeper", table_name);
+
+        assert(!ddl_worker || !ddl_worker->isCurrentlyActive());
+
+        const String statement = getObjectDefinitionFromCreateQuery(getCreateTableQuery(table_name, local_context));
+        const String table_metadata_zk_path = zookeeper_path + "/metadata/" + escapeForFileName(table_name);
+        txn->addOp(zkutil::makeCreateRequest(table_metadata_zk_path, statement, zkutil::CreateMode::Persistent));
+
+        tables_digest += DB::getMetadataHash(table_name, statement);
+    }
+
+    txn->addOp(zkutil::makeSetRequest(metadata_node_path, "", entity_name_stat.version));
+    txn->addOp(zkutil::makeSetRequest(replica_path + "/digest", toString(tables_digest), -1));
+    {
+        std::lock_guard lock{metadata_mutex};
+        tables_metadata_digest = tables_digest;
+        assert(checkDigestValid(local_context));
```

If the digest is not valid, should we throw an error instead?